core/src/autogluon/core/trainer/abstract_trainer.py

```diff
@@ -45,18 +49,213 @@
+        was trained with the name
+        """
+        return self.get_model_names() + list(self._extra_banned_names)
+
+    @property
+    def path_root(self) -> str:
+        """directory containing learner.pkl"""
+        return os.path.dirname(self.path)
+
+    @property
+    def path_utils(self) -> str:
+        return os.path.join(self.path_root, "utils")
+
+    @property
+    def path_data(self) -> str:
+        return os.path.join(self.path_utils, "data")
+
+    def set_contexts(self, path_context: str) -> None:
+        self.path = self.create_contexts(path_context)
+
+    def create_contexts(self, path_context: str) -> str:
+        path = path_context
+        return path
+
+    def save_model(self, model: AbstractModel, **kwargs) -> None:
+        model.save()
+        if not self.low_memory:
+            self.models[model.name] = model
+
+    def get_models_attribute_dict(self, attribute: str, models: list | None = None) -> dict[str, Any]:
+        """Returns dictionary of model name -> attribute value for the provided attribute.
+        """
+        models_attribute_dict = nx.get_node_attributes(self.model_graph, attribute)
+        if models is not None:
+            model_names = []
+            for model in models:
+                if not isinstance(model, str):
+                    model = model.name
+                model_names.append(model)
+            if attribute == "path":
+                models_attribute_dict = {key: os.path.join(*val) for key, val in models_attribute_dict.items() if key in model_names}
+            else:
+                models_attribute_dict = {key: val for key, val in models_attribute_dict.items() if key in model_names}
+        return models_attribute_dict
+
+    def get_model_attribute(self, model: AbstractModel | str, attribute: str, **kwargs) -> Any:
+        """Return model attribute value.
+        If `default` is specified, return default value if attribute does not exist.
+        If `default` is not specified, raise ValueError if attribute does not exist.
+        """
+        if not isinstance(model, str):
+            model = model.name
+        if model not in self.model_graph.nodes:
+            raise ValueError(f"Model does not exist: (model={model})")
+        if attribute not in self.model_graph.nodes[model]:
+            if "default" in kwargs:
+                return kwargs["default"]
+            else:
+                raise ValueError(f"Model does not contain attribute: (model={model}, attribute={attribute})")
+        if attribute == "path":
+            return os.path.join(*self.model_graph.nodes[model][attribute])
+        return self.model_graph.nodes[model][attribute]
+
+    def set_model_attribute(self, model, attribute: str, val):
+        if not isinstance(model, str):
+            model = model.name
+        self.model_graph.nodes[model][attribute] = val
+
+    def get_minimum_model_set(self, model, include_self=True) -> list:
```

Comment:
Nit: Some methods are missing type hints.

Reply:
I was really hoping that by specializing in Python instead of C++ I will be able to avoid learning what a generic is... welp